pkg/autoupdate/tuf/autoupdate.go

```diff
@@ -0,0 +1,231 @@
+		ta.logger = log.With(logger, "component", "tuf_autoupdater")
+	}
+}
+
+func WithChannel(channel string) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.channel = channel
+	}
+}
+
+func WithUpdateCheckInterval(checkInterval time.Duration) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.checkInterval = checkInterval
+	}
+}
+
+func NewTufAutoupdater(metadataUrl, binary, rootDirectory string, metadataHttpClient *http.Client, opts ...TufAutoupdaterOption) (*TufAutoupdater, error) {
+	ta := &TufAutoupdater{
+		binary:          binary,
+		operatingSystem: runtime.GOOS,
+		channel:         DefaultChannel,
+		interrupt:       make(chan struct{}),
+		checkInterval:   60 * time.Second,
+		errorCounter:    make([]int64, 0), // For now, the error counter is a simple list of timestamps when errors occurred
+		logger:          log.NewNopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(ta)
+	}
+
+	var err error
+	ta.metadataClient, err = initMetadataClient(binary, rootDirectory, metadataUrl, metadataHttpClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not init metadata client: %w", err)
+	}
+
+	return ta, nil
+}
+
+func initMetadataClient(binary, rootDirectory, metadataUrl string, metadataHttpClient *http.Client) (*client.Client, error) {
+	// Set up the local TUF directory for our TUF client -- a dev repo, to be replaced once we move to production
+	localTufDirectory := filepath.Join(rootDirectory, fmt.Sprintf("%s-tuf-dev", binary))
+	if err := os.MkdirAll(localTufDirectory, 0750); err != nil {
+		return nil, fmt.Errorf("could not make local TUF directory %s: %w", localTufDirectory, err)
+	}
+
+	// Set up our local store i.e. point to the directory in our filesystem
+	localStore, err := filejsonstore.NewFileJSONStore(localTufDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize local TUF store: %w", err)
+	}
+
+	// Set up our remote store i.e. tuf-devel.kolide.com
+	remoteOpts := client.HTTPRemoteOptions{
+		MetadataPath: "/repository",
+	}
+	remoteStore, err := client.HTTPRemoteStore(metadataUrl, &remoteOpts, metadataHttpClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize remote TUF store: %w", err)
+	}
+
+	metadataClient := client.NewClient(localStore, remoteStore)
+	if err := metadataClient.Init(rootJson); err != nil {
+		return nil, fmt.Errorf("failed to initialize TUF client with root JSON: %w", err)
+	}
+
+	return metadataClient, nil
+}
+
+func (ta *TufAutoupdater) Execute() (err error) {
+	checkTicker := time.NewTicker(ta.checkInterval)
+	errorCheckTicker := time.NewTicker(1 * time.Hour)
+	cleanupTicker := time.NewTicker(12 * time.Hour)
+
+	for {
+		select {
+		case <-checkTicker.C:
+			if err := ta.checkForUpdate(); err != nil {
+				ta.lock.Lock()
+				ta.errorCounter = append(ta.errorCounter, time.Now().Unix())
+				ta.lock.Unlock()
+
+				level.Debug(ta.logger).Log("msg", "error checking for update", "err", err)
+			}
+		case <-errorCheckTicker.C:
```

Comment:
Why are we doing the error count check and error clean up thing? Why don't we just log an error and move on? We are logging it in the checkTime.C block above.

Reply:
In the next PR, I want to expose the error count in a launcher table, kind of similar to `kolide_launcher_osquery_instance_history` -- I can add a comment about what it's for!

Reply:
Added a comment on the error counter!